File: polyglot/python/ais_stream_parser.py
"""
polyglot/python/ais_stream_parser.py

AIS Stream Parser for Maritime Situational Awareness / Defensive OSINT

Parses NMEA 0183 and AIS binary streams, extracts vessel tracks, detects
transponder-gap anomalies (sudden position jumps, speed spikes, course whiplash).

Usage:
    python -m polyglot.python.ais_stream_parser <input.nmea> [--binary]
"""

import argparse
import csv
import io
import math
import os
import sys
from collections import deque
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any, Callable, Deque, Iterator, Optional, TextIO

# =============================================================================
# CONSTANTS & CONFIGURATION
# =============================================================================

NMEA_SENTENCE_LENGTH_LIMIT = 256
DEFAULT_BUFFER_SIZE = 4096
MIN_TRACK_POINTS_FOR_GAP = 3
GAP_THRESHOLD_METERS = 100.0
SPEED_SPIKE_MULTIPLIER = 3.0
COURSE_WHIPLASH_DEGREES = 45.0

# NMEA Sentence Types (ISO 13628-5 / IEC 61993-2)
NMEA_SENTENCE_TYPES: dict[str, str] = {
    "GGA": "Global Positioning System Fix Data",
    "RMB": "Recommended Minimum Navigation Information",
    "VTG": "Course over Ground (COG)",
    "VWR": "Water Speed Over Ground",
    "HDM": "Heading Magnetic",
    "HDA": "Heading True",
    "HDG": "True Heading",
    "XTE": "Cross Track Error",
    "RMA": "Recommended Minimum Navigation Information - Aided",
    "RMC": "Recommended Minimum Navigation Information (Position)",
    "ZDA": "Zulu Time and Date",
    "VDM": "Class B Data Message",
    "VDO": "Class B Data Message - Outbound",
    "VDN": "Class B Data Message - NMEA 0183 Format",
}

# AIS Binary Frame Types (ISO 19765)
AIS_FRAME_TYPES: dict[int, str] = {
    1: "Position Report (SOTDMA)",
    2: "Position Report (SOTDMA/ITDMA)",
    3: "Position Report (ITDMA)",
    4: "SAR Aircraft Position",
    5: "Static Data & Voyage Data",
    6: "Dynamic Data",
    7: "Rapid Update",
    8: "Extended Class B",
    9: "Class A ARQ/FR",
    10: "Position Report (ARQ)",
    11: "Static & Voyage Data (ARQ)",
    12: "Dynamic Data (ARQ)",
    13: "Rapid Update (ARQ)",
    14: "Extended Class B (ARQ)",
    15: "Position Report (FR)",
    16: "Static & Voyage Data (FR)",
    17: "Dynamic Data (FR)",
    18: "Rapid Update (FR)",
    19: "Extended Class B (FR)",
}

# =============================================================================
# DATA MODELS
# =============================================================================

class NMEASentenceType(Enum):
    GGA = "GGA"
    RMB = "RMB"
    VTG = "VTG"
    VWR = "VWR"
    HDM = "HDM"
    HDA = "HDA"
    HDG = "HDG"
    XTE = "XTE"
    RMA = "RMA"
    RMC = "RMC"
    ZDA = "ZDA"
    VDM = "VDM"
    VDO = "VDO"
    VDN = "VDN"

@dataclass(frozen=True)
class NMEASentence:
    """Raw parsed NMEA sentence."""
    raw: str
    type: NMEASentenceType
    checksum_valid: bool
    fields: list[str]  # Original field values (before splitting)
    
    @classmethod
    def parse(cls, raw: str) -> Optional["NMEASentence"]:
        """Parse a single NMEA sentence."""
        if not raw or len(raw) > NMEA_SENTENCE_LENGTH_LIMIT:
            return None
        
        fields = raw.split(",")
        
        # Validate checksum (last 2 chars must be *XX)
        if len(fields) < 3 or not fields[-1].startswith("*"):
            return None
        
        try:
            checksum_chars = fields[-1][1:]
            expected_checksum = sum(ord(c) for c in raw[:-3]) & 0xFF
            actual_checksum = int(checksum_chars, 16)
            checksum_valid = (expected_checksum == actual_checksum)
        except ValueError:
            return None
        
        # Extract sentence type
        if len(fields) < 2:
            return None
        
        try:
            type_val = fields[0].upper()
            type_enum = NMEASentenceType(type_val)
        except (ValueError, KeyError):
            type_enum = NMEASentenceType.GGA  # Default fallback
            
        return cls(raw=raw, type=type_enum, checksum_valid=checksum_valid, fields=fields)

@dataclass(frozen=True)
class VesselPosition:
    """Single vessel position fix."""
    timestamp: datetime
    lat: float  # Degrees N/S (positive = North)
    lon: float  # Degrees E/W (positive = East)
    speed_over_ground: Optional[float] = None  # Knots
    course_over_ground: Optional[float] = None  # Degrees True
    true_heading: Optional[float] = None  # Degrees True
    mag_heading: Optional[float] = None  # Degrees Magnetic
    water_speed: Optional[float] = None  # Knots
    cross_track_error: Optional[float] = None  # Meters (signed)
    quality_flags: int = 0  # GGA quality indicator
    
    def __post_init__(self):
        if self.quality_flags & 1:
            self.lat = -self.lat  # Negative = South
        
        if self.quality_flags & 2:
            self.lon = -self.lon  # Negative = West

@dataclass(frozen=True)
class AISMessage:
    """Parsed AIS binary message."""
    frame_type: int
    mmsi: int
    timestamp: datetime
    lat: float
    lon: float
    sog: Optional[float] = None  # Speed over ground (knots)
    cog: Optional[float] = None   # Course over ground (degrees)
    hdg: Optional[float] = None   # Heading (degrees)
    raim: bool = False           # Receiver in autonomous mode
    raim_quality: int = 0        # RAIM quality indicator
    
@dataclass(frozen=True)
class TrackPoint:
    """Aggregated track point with derived metrics."""
    vessel_id: str
    timestamp: datetime
    lat: float
    lon: float
    sog: Optional[float] = None
    cog: Optional[float] = None
    hdg: Optional[float] = None
    
    # Derived anomaly scores (0.0 - 1.0, higher = worse)
    gap_score: float = 0.0
    speed_anomaly_score: float = 0.0
    course_whiplash_score: float = 0.0
    total_anomaly_score: float = 0.0
    
    # Previous values for comparison (set during stream processing)
    prev_lat: Optional[float] = None
    prev_lon: Optional[float] = None
    prev_sog: Optional[float] = None
    prev_cog: Optional[float] = None
    prev_hdg: Optional[float] = None
    
    @property
    def is_anomalous(self) -> bool:
        return self.total_anomaly_score > 0.5

@dataclass(frozen=True)
class StreamState:
    """Per-vessel state maintained during stream processing."""
    vessel_id: str
    points: Deque[TrackPoint] = field(default_factory=deque)
    
    # Running statistics for anomaly detection
    avg_sog: float = 0.0
    sog_stddev: float = 0.0
    last_gap_score: float = 0.0
    
    @property
    def count(self) -> int:
        return len(self.points)

# =============================================================================
# UTILITY FUNCTIONS
# =============================================================================

def degrees_to_radians(degrees: float) -> float:
    """Convert degrees to radians."""
    return math.radians(degrees)

def radians_to_degrees(radians: float) -> float:
    """Convert radians to degrees."""
    return math.degrees(radians)

def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate great-circle distance between two points in meters.
    
    Args:
        lat1, lon1: First point (degrees)
        lat2, lon2: Second point (degrees)
        
    Returns:
        Distance in nautical miles (multiply by 1852 for meters)
    """
    R = 3440.065  # Earth radius in nautical miles
    
    dlat = degrees_to_radians(lat2 - lat1)
    dlon = degrees_to_radians(lon2 - lon1)
    
    a = (math.sin(dlat / 2) ** 2 + 
         math.cos(degrees_to_radians(lat1)) * math.cos(degrees_to_radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c

def calculate_track_speed(lat1: float, lon1: float, 
                         lat2: float, lon2: float,
                         time_diff_seconds: float) -> Optional[float]:
    """Calculate speed over ground from two track points.
    
    Args:
        (lat1, lon1), (lat2, lon2): Two consecutive positions (degrees)
        time_diff_seconds: Time difference between fixes
        
    Returns:
        Speed in knots, or None if time_diff is zero or negative
    """
    if time_diff_seconds <= 0:
        return None
    
    distance_nm = haversine_distance(lat1, lon1, lat2, lon2)
    speed_knots = (distance_nm / time_diff_seconds) * 60.0
    return max(0.0, speed_knots)

def calculate_course_change(cog1: Optional[float], cog2: Optional[float]) -> float:
    """Calculate absolute course change between two fixes.
    
    Args:
        cog1, cog2: Course over ground (degrees True), or None
        
    Returns:
        Absolute course change in degrees [0, 360)
    """
    if cog1 is None or cog2 is None:
        return 0.0
    
    diff = abs(cog2 - cog1) % 360
    return min(diff, 360 - diff)

def calculate_heading_change(hdg1: Optional[float], hdg2: Optional[float]) -> float:
    """Calculate absolute heading change between two fixes."""
    if hdg1 is None or hdg2 is None:
        return 0.0
    
    diff = abs(hdg2 - hdg1) % 360
    return min(diff, 360 - diff)

def calculate_anomaly_score(base_score: float, 
                           multiplier: float,
                           threshold: float) -> float:
    """Calculate anomaly score with saturation at 1.0."""
    raw = base_score * multiplier
    return min(1.0, max(0.0, raw))

# =============================================================================
# NMEA STREAM PARSER
# =============================================================================

class NMESentenceParser:
    """Parse and buffer NMEA sentences from a stream."""
    
    def __init__(self, buffer_size: int = DEFAULT_BUFFER_SIZE):
        self.buffer_size = buffer_size
        self._buffer: Deque[str] = deque(maxlen=buffer_size)
        
    def feed(self, data: str) -> Iterator[NMEASentence]:
        """Feed raw NMEA data and yield parsed sentences."""
        # Split into individual sentences (handle multi-line records)
        lines = data.strip().split("\n")
        
        for line in lines:
            line = line.strip()
            if not line or line[0] == "!":  # Skip empty/comment lines
                continue
                
            sentence = NMEASentence.parse(line)
            if sentence:
                self._buffer.append(sentence)
                yield sentence
    
    def flush(self) -> Iterator[NMEASentence]:
        """Yield any remaining buffered sentences."""
        for sentence in self._buffer:
            yield sentence
        self._buffer.clear()

# =============================================================================
# AIS BINARY PARSER (ISO 19765 / IEC 61993-2)
# =============================================================================

class AISBinaryParser:
    """Parse raw AIS binary frames from NMEA VDM/VDO/VDN sentences."""
    
    # AIS Frame Header Format (bits):
    #   F0-F1: Frame Type (4 bits, 0x00-0x0F)
    #   F2: RAIM flag (1 bit)
    #   F3: Reserved (1 bit)
    #   F4-F5: Message ID (2 bits)
    
    @classmethod
    def parse_vdm(cls, vdm: str) -> Optional[AISMessage]:
        """Parse a VDM (Class A) sentence.
        
        VDM format: <FrameType><RAIMFlag><MsgID><Data>
        Example: 1023456789ABCDEF...
        """
        if not vdm or len(vdm) < 4:
            return None
        
        try:
            # Extract header fields
            frame_type = int(vdm[0]) & 0x0F
            raim_flag = (int(vdm[1]) >> 6) & 0x01
            msg_id = (int(vdm[2]) >> 4) & 0x03
            
            # Validate frame type
            if frame_type not in AIS_FRAME_TYPES:
                return None
            
            # Parse message data (simplified - full parsing requires ISO 19765 spec)
            # For demo purposes, extract basic fields from hex payload
            payload = vdm[4:]
            
            # Decode latitude and longitude (20 bits each, little-endian)
            lat_bits = int(payload[:4], 16) if len(payload) >= 4 else 0
            lon_bits = int(payload[4:8], 16) if len(payload) >= 8 else 0
            
            # Convert to degrees (20-bit fixed point, 1/100000 scale)
            lat_degrees = (lat_bits / 100000.0) * 60.0 - 90.0 if lat_bits > 3145728 else (lat_bits / 100000.0) * 60.0 - 90.0
            lon_degrees = (lon_bits / 100000.0) * 60.0 - 180.0 if lon_bits > 3145728 else (lon_bits / 100000.0) * 60.0 - 180.0
            
            # Adjust for hemisphere
            if lat_degrees < 0:
                lat_degrees = 360.0 + lat_degrees
            
            return AISMessage(
                frame_type=frame_type,
                mmsi=int(vdm[2]),
                timestamp=datetime.now(timezone.utc),
                lat=lat_degrees,
                lon=lon_degrees,
                raim=bool(raim_flag)
            )
            
        except (ValueError, IndexError):
            return None

# =============================================================================
# TRACK PROCESSOR & ANOMALY DETECTOR
# =============================================================================

class TrackProcessor:
    """Process vessel tracks and detect anomalies."""
    
    def __init__(self, 
                 gap_threshold_meters: float = GAP_THRESHOLD_METERS,
                 speed_spike_multiplier: float = SPEED_SPIKE_MULTIPLIER,
                 course_whiplash_degrees: float = COURSE_WHIPLASH_DEGREES):
        self.gap_threshold_meters = gap_threshold_meters
        self.speed_spike_multiplier = speed_spike_multiplier
        self.course_whiplash_degrees = course_whiplash_degrees
        
    def process_point(self, point: TrackPoint, 
                      state: StreamState) -> tuple[TrackPoint, StreamState]:
        """Process a single track point and update anomaly scores."""
        
        # Calculate gap score (position jump anomaly)
        if len(state.points) >= MIN_TRACK_POINTS_FOR_GAP:
            prev = state.points[-1]
            
            distance_nm = haversine_distance(
                prev.lat, prev.lon, 
                point.lat, point.lon
            )
            distance_meters = distance_nm * 1852.0
            
            if distance_meters > self.gap_threshold_meters:
                # Exponential decay with gap size
                raw_gap_score = min(1.0, (distance_meters

File: polyglot/go/ais_stream_parser.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

// AISMessage represents a parsed NMEA 0183 position fix message
type AISMessage struct {
	Timestamp time.Time
	Lat       float64 // Degrees, North positive
	Lon       float64 // Degrees, East positive
	Speed     float64 // Knots (SOG)
	Course    float64 // Degrees true
	Heading   float64 // Degrees magnetic
}

// GapAnomaly represents a detected gap between consecutive positions
type GapAnomaly struct {
	Timestamp       time.Time
	PreviousLat     float64
	PreviousLon     float64
	CurrentLat      float64
	CurrentLon      float64
	DurationMinutes float64
	DistanceNM      float64
	Score           int
}

// GapConfig holds configuration for gap detection
type GapConfig struct {
	MaxNormalDeltaMin  float64 // Maximum normal time delta (minutes) - typically 1-2 min
	MaxNormalDistNMI   float64 // Maximum normal distance between fixes (nautical miles)
	ScoreThreshold     int    // Minimum score to flag as anomaly
}

// Default configuration values
var DefaultConfig = GapConfig{
	MaxNormalDeltaMin:  2.0,
	MaxNormalDistNMI:   5.0,
	ScoreThreshold:     3,
}

// CalculateDistanceNM computes distance between two latitude/longitude points in nautical miles
func CalculateDistanceNM(lat1, lon1, lat2, lon2 float64) float64 {
	latDiff := math.Abs(lat2 - lat1) * 60.0 // Convert to degrees
	lonDiff := math.Abs(lon2 - lon1) * 60.0
	
	// Haversine formula for more accuracy at larger distances
	const earthRadiusNM = 3440.065
	a := 0.5 * (math.Sin(latDiff/2)*math.Sin(latDiff/2) + 
		math.Cos(lat1*math.Pi/180.0)*math.Cos(lat2*math.Pi/180.0)*
			math.Sin(lonDiff/2)*math.Sin(lonDiff/2))
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	
	return earthRadiusNM * c
}

// ParseNMEATimestamp converts NMEA timestamp to time.Time
func ParseNMEATimestamp(timestamp string) (time.Time, error) {
	// Format: HHMMSS.SSS (e.g., 123456.789)
	if len(timestamp) < 6 {
		return time.Time{}, fmt.Errorf("invalid timestamp length: %s", timestamp)
	}
	
	hour, err := strconv.Atoi(timestamp[:2])
	if err != nil {
		return time.Time{}, fmt.Errorf("invalid hour in timestamp: %s", timestamp)
	}
	
	minute, err := strconv.Atoi(timestamp[2:4])
	if err != nil {
		return time.Time{}, fmt.Errorf("invalid minute in timestamp: %s", timestamp)
	}
	
	second, err := strconv.ParseFloat(timestamp[4:], 64)
	if err != nil {
		return time.Time{}, fmt.Errorf("invalid second in timestamp: %s", timestamp)
	}
	
	return time.Date(2024, 1, 1, hour, minute, int(second), 0, time.UTC), nil
}

// ParseAISMessage parses a NMEA position fix message into AISMessage struct
func ParseAISMessage(nmea string) (*AISMessage, error) {
	// Expected format: $GPRMC,hhmmss.ss,A,dd.mmnnnnn,x.xxxxxxx,N,E,...*checksum
	parts := strings.SplitN(nmea, ",", 10)
	if len(parts) < 8 {
		return nil, fmt.Errorf("insufficient fields in NMEA message: %s", nmea)
	}
	
	timestamp, err := ParseNMEATimestamp(parts[1])
	if err != nil {
		return nil, err
	}
	
	var lat, lon float64
	status := parts[2] // A = active, V = void
	
	if status == "A" && len(parts) >= 7 {
		// Parse latitude: dd.mmnnnnn (e.g., 37.8012345 -> 37.8012345°)
		latStr := parts[3]
		if strings.Contains(latStr, ".") {
			lat, err = strconv.ParseFloat(latStr[:strings.Index(latStr, ".")+1], 64)
			if err != nil {
				return nil, fmt.Errorf("invalid latitude: %s", latStr)
			}
		} else {
			// Handle ddmm.mmm format (multiply by 60)
			lat, err = strconv.ParseFloat(latStr, 64)
			if err != nil {
				return nil, fmt.Errorf("invalid latitude: %s", latStr)
			}
			lat *= 60.0
		}
		
		// Parse longitude: x.xxxxxxx (e.g., 122.456789 -> 122.456789°)
		lonStr := parts[4]
		if strings.Contains(lonStr, ".") {
			lon, err = strconv.ParseFloat(lonStr[:strings.Index(lonStr, ".")+1], 64)
			if err != nil {
				return nil, fmt.Errorf("invalid longitude: %s", lonStr)
			}
		} else {
			lon, err = strconv.ParseFloat(lonStr, 64)
			if err != nil {
				return nil, fmt.Errorf("invalid longitude: %s", lonStr)
			}
			lon *= 60.0
		}
		
		var speed float64
		if len(parts) >= 9 && parts[8] != "" {
			speed, err = strconv.ParseFloat(parts[8], 64)
			if err != nil {
				return nil, fmt.Errorf("invalid speed: %s", parts[8])
			}
		}
		
		var course float64
		if len(parts) >= 10 && parts[9] != "" {
			course, err = strconv.ParseFloat(parts[9], 64)
			if err != nil {
				return nil, fmt.Errorf("invalid course: %s", parts[9])
			}
		}
		
		var heading float64
		if len(parts) >= 11 && parts[10] != "" {
			heading, err = strconv.ParseFloat(parts[10], 64)
			if err != nil {
				return nil, fmt.Errorf("invalid heading: %s", parts[10])
			}
		}
		
		return &AISMessage{
			Timestamp: timestamp,
			Lat:       lat,
			Lon:       lon,
			Speed:     speed,
			Course:    course,
			Heading:   heading,
		}, nil
	}
	
	return nil, fmt.Errorf("inactive or void AIS message")
}

// AnalyzeAISStream processes a slice of NMEA strings and detects gaps
func AnalyzeAISStream(messages []string) ([]GapAnomaly, error) {
	var anomalies []GapAnomaly
	var prev *AISMessage
	
	for _, nmea := range messages {
		parsed, err := ParseAISMessage(nmea)
		if err != nil {
			continue // Skip malformed messages
		}
		
		if parsed == nil || !parsed.Timestamp.IsZero() {
			continue
		}
		
		if prev != nil && !prev.Timestamp.IsZero() {
			duration := float64(parsed.Timestamp.Sub(prev.Timestamp).Minutes())
			distance := CalculateDistanceNM(
				prev.Lat, prev.Lon, parsed.Lat, parsed.Lon)
			
			// Determine if this is a gap anomaly
			if duration > DefaultConfig.MaxNormalDeltaMin || 
			   distance > DefaultConfig.MaxNormalDistNMI {
				
				score := calculateGapScore(duration, distance)
				
				if score >= DefaultConfig.ScoreThreshold {
					anomalies = append(anomalies, GapAnomaly{
						Timestamp:       parsed.Timestamp,
						PreviousLat:     prev.Lat,
						PreviousLon:     prev.Lon,
						CurrentLat:      parsed.Lat,
						CurrentLon:      parsed.Lon,
						DurationMinutes: duration,
						DistanceNM:      distance,
						Score:           score,
					})
				}
			}
		}
		
		prev = parsed
	}
	
	return anomalies, nil
}

// calculateGapScore computes an anomaly score based on gap characteristics
func calculateGapScore(durationMinutes, distanceNM float64) int {
	var score int
	
	// Base score from duration
	if durationMinutes < 5 {
		score += 2 // Minor gap
	} else if durationMinutes < 30 {
		score += 5 // Moderate gap
	} else if durationMinutes < 60 {
		score += 8 // Significant gap
	} else {
		score += 12 // Major gap
	}
	
	// Add penalty for distance traveled during gap (unexpected jump)
	if distanceNM > DefaultConfig.MaxNormalDistNMI*3 {
		score += 4 // Large spatial jump
	} else if distanceNM > DefaultConfig.MaxNormalDistNMI {
		score += 2 // Moderate spatial jump
	}
	
	// Cap score at reasonable maximum
	if score > 20 {
		score = 20
	}
	
	return score
}

// FormatAnomalyOutput formats a GapAnomaly for display
func FormatAnomalyOutput(anomaly GapAnomaly) string {
	return fmt.Sprintf(
		"Gap Anomaly [Score: %d/20]\n"+
			"  Time: %s → %s\n"+
			"  Duration: %.1f min | Distance: %.3f NM\n"+
			"  Previous: %.4f°N, %.4f°E\n"+
			"  Current:  %.4f°N, %.4f°E",
		anomaly.Score,
		anomaly.Timestamp.Format("15:04:05"),
		anomaly.Timestamp.Add(time.Duration(anomaly.DurationMinutes*60)).Format("15:04:05"),
		anomaly.DurationMinutes,
		anomaly.DistanceNM,
		anomaly.PreviousLat, anomaly.PreviousLon,
		anomaly.CurrentLat, anomaly.CurrentLon,
	)
}

// RunDemo demonstrates the AIS stream parser with sample data
func RunDemo() {
	fmt.Println("=== AIS Stream Parser Demo ===\n")
	
	// Sample NMEA messages simulating a vessel track with gaps
	sampleMessages := []string{
		"$GPRMC,123456.78,A,37.8012345,N,122.456789,W,12.34,145.67,0.00,A,37.8012345,N,122.456789,W*6A",
		"$GPRMC,123457.12,A,37.8014567,N,122.4569012,W,12.35,145.68,0.00,A,37.8014567,N,122.4569012,W*6B",
		"$GPRMC,123458.45,A,37.8016789,N,122.4570123,W,12.36,145.69,0.00,A,37.8016789,N,122.4570123,W*6C",
		// Normal progression...
		"$GPRMC,123460.12,A,37.8023456,N,122.4575678,W,12.38,145.71,0.00,A,37.8023456,N,122.4575678,W*6D",
		"$GPRMC,123461.34,A,37.8025678,N,122.4577890,W,12.39,145.72,0.00,A,37.8025678,N,122.4577890,W*6E",
		// Gap introduced - vessel jumps forward in time
		"$GPRMC,123500.56,A,37.8056789,N,122.4612345,W,12.45,145.80,0.00,A,37.8056789,N,122.4612345,W*6F",
		"$GPRMC,123501.78,A,37.8058901,N,122.4614567,W,12.46,145.81,0.00,A,37.8058901,N,122.4614567,W*6G",
	}
	
	fmt.Println("Processing sample AIS stream...")
	anomalies, err := AnalyzeAISStream(sampleMessages)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	
	if len(anomalies) == 0 {
		fmt.Println("No significant gaps detected.")
	} else {
		fmt.Printf("\nDetected %d gap anomaly(ies):\n\n", len(anomalies))
		
		for i, a := range anomalies {
			fmt.Printf("--- Anomaly #%d ---\n", i+1)
			fmt.Println(FormatAnomalyOutput(a))
			
			// Severity classification
			if a.Score >= 15 {
				fmt.Println("Severity: HIGH - Investigate vessel trajectory")
			} else if a.Score >= 8 {
				fmt.Println("Severity: MEDIUM - Monitor closely")
			} else {
				fmt.Println("Severity: LOW - Normal variation")
			}
		}
		
		// Summary statistics
		totalDuration := float64(0)
		totalDistance := float64(0)
		for _, a := range anomalies {
			totalDuration += a.DurationMinutes
			totalDistance += a.DistanceNM
		}
		
		fmt.Printf("\n=== Summary ===\n")
		fmt.Printf("Total gap time: %.1f minutes\n", totalDuration)
		fmt.Printf("Total distance jumped: %.3f NM\n", totalDistance)
		fmt.Printf("Average score: %.1f/20\n", float64(anomalies[0].Score)) // Simplified for demo
	}
	
	// Demonstrate with a more obvious gap scenario
	fmt.Println("\n=== Test with Obvious Gap ===\n")
	
	obviousGapMessages := []string{
		"$GPRMC,123456.00,A,37.8000000,N,122.4500000,W,10.00,000.00,0.00,A*6A",
		"$GPRMC,123457.00,A,37.8001000,N,122.4501000,W,10.01,000.01,0.00,A*6B",
		"$GPRMC,123500.00,A,37.8050000,N,122.4550000,W,10.50,000.50,0.00,A*6